src/tests/suite/daos_dedup.c

```diff
@@ -0,0 +1,361 @@
+}
+
+static void
+setup_keys(struct dedup_test_ctx *ctx)
+{
+	iov_alloc_str(&ctx->dkey, "dkey");
+	iov_alloc_str(&ctx->update_iod.iod_name, "akey");
+}
+
+static void
+setup_as_array(struct dedup_test_ctx *ctx)
+{
+	ctx->recx[0].rx_idx = 0;
+	ctx->recx[0].rx_nr = daos_sgl_buf_size(&ctx->update_sgl);
+	ctx->update_iod.iod_size = 1;
+	ctx->update_iod.iod_nr	= 1;
+	ctx->update_iod.iod_recxs = &ctx->recx[0];
+	ctx->update_iod.iod_type  = DAOS_IOD_ARRAY;
+}
+
+static void
+setup_as_single_value(struct dedup_test_ctx *ctx)
+{
+	ctx->update_iod.iod_nr	= 1;
+	ctx->update_iod.iod_size = daos_sgl_buf_size(&ctx->update_sgl);
+	ctx->update_iod.iod_recxs = NULL;
+	ctx->update_iod.iod_type  = DAOS_IOD_SINGLE;
+}
+
+static void
+setup_fetch_iod(struct dedup_test_ctx *ctx)
+{
+	ctx->fetch_iod.iod_name = ctx->update_iod.iod_name;
+	ctx->fetch_iod.iod_size = ctx->update_iod.iod_size;
+	ctx->fetch_iod.iod_recxs = ctx->update_iod.iod_recxs;
+	ctx->fetch_iod.iod_nr = ctx->update_iod.iod_nr;
+	ctx->fetch_iod.iod_type = ctx->update_iod.iod_type;
+}
+
+static void
+setup_cont_obj(struct dedup_test_ctx *ctx,
+	       uint32_t csum_prop_type,
+	       daos_oclass_id_t oclass,
+	       uint32_t dedup_type,
+	       uint32_t dedup_threshold_setting)
+{
+	daos_prop_t *props = NULL;
+	int		 rc;
+	daos_size_t	 data_len;
+	daos_size_t	 dedup_threshold;
+
+
+	/** calc threshold based on data size and setting */
+	data_len = daos_sgl_buf_size(&ctx->update_sgl);
+	dedup_threshold = dedup_threshold_setting == THRESHOLD_GREATER_THAN_DATA
+			  ? data_len + 10 : data_len - 10;
+
+	uuid_generate(ctx->uuid);
+
+	props = daos_prop_alloc(3);
+	assert_non_null(props);
+	props->dpp_entries[0].dpe_type = DAOS_PROP_CO_CSUM;
+	props->dpp_entries[0].dpe_val = csum_prop_type;
+	props->dpp_entries[1].dpe_type = DAOS_PROP_CO_DEDUP;
+	props->dpp_entries[1].dpe_val = dedup_type;
+	props->dpp_entries[2].dpe_type = DAOS_PROP_CO_DEDUP_THRESHOLD;
+	props->dpp_entries[2].dpe_val = dedup_threshold;
+
+	rc = daos_cont_create(ctx->poh, ctx->uuid, props, NULL);
+	assert_int_equal(0, rc);
+	daos_prop_free(props);
+
+	rc = daos_cont_open(ctx->poh, ctx->uuid, DAOS_COO_RW,
+			    &ctx->coh, &ctx->info, NULL);
+	assert_int_equal(0, rc);
+
+	ctx->oid = dts_oid_gen(oclass, 0, 1);
+	rc = daos_obj_open(ctx->coh, ctx->oid, 0, &ctx->oh, NULL);
+	assert_int_equal(0, rc);
+}
+
+static void
+setup_context(struct dedup_test_ctx *ctx, test_arg_t *state,
+	      uint32_t iod_type,
+	      uint32_t csum_prop_type,
+	      daos_oclass_id_t oclass,
+	      uint32_t dedup_type,
+	      enum THRESHOLD_SETTING dedup_threshold_setting)
+{
+	setup_from_test_args(ctx, state);
+	setup_keys(ctx);
+	setup_sgl(ctx);
+
+	if (iod_type == DAOS_IOD_ARRAY)
+		setup_as_array(ctx);
+	else if (iod_type == DAOS_IOD_SINGLE)
+		setup_as_single_value(ctx);
+	else
+		fail_msg("Invalid iod_type: %d\n", iod_type);
+
+	setup_fetch_iod(ctx);
+
+	setup_cont_obj(ctx, csum_prop_type, oclass,
+		       dedup_type, dedup_threshold_setting);
```

Comment:
(style)  'dedup' may be misspelled - perhaps 'dedupe'?
(style)  'dedup' may be misspelled - perhaps 'dedupe'?